File: python-selenium-pytest-bdd/src/saucedemo_uitests/pages/checkout_information_page.py
from selenium.webdriver.common.by import By

from saucedemo_uitests.pages.base_page import BasePage, Locator
from saucedemo_uitests.pages.checkout_overview_page import CheckoutOverviewPage
from saucedemo_uitests.testdata.checkout_info import CheckoutInfo

_FIRST_NAME_INPUT: Locator = (By.CSS_SELECTOR, "[data-test='firstName']")
_LAST_NAME_INPUT: Locator = (By.CSS_SELECTOR, "[data-test='lastName']")
_POSTAL_CODE_INPUT: Locator = (By.CSS_SELECTOR, "[data-test='postalCode']")
_CONTINUE_BUTTON: Locator = (By.CSS_SELECTOR, "[data-test='continue']")
_ERROR_BANNER: Locator = (By.CSS_SELECTOR, "[data-test='error']")


class CheckoutInformationPage(BasePage):
    def fill_and_continue(self, info: CheckoutInfo) -> CheckoutOverviewPage:
        self.type_text(_FIRST_NAME_INPUT, info.first_name)
        self.type_text(_LAST_NAME_INPUT, info.last_name)
        self.type_text(_POSTAL_CODE_INPUT, info.postal_code)
        self.click(_CONTINUE_BUTTON)
        return CheckoutOverviewPage(self.driver)

    def submit_without_postal_code(self, info: CheckoutInfo) -> None:
        """Leaves the postal code blank and submits, for the "postal code is required" negative
        scenario."""
        self.type_text(_FIRST_NAME_INPUT, info.first_name)
        self.type_text(_LAST_NAME_INPUT, info.last_name)
        self.click(_CONTINUE_BUTTON)

    def get_error_message(self) -> str:
        return self.text_of(_ERROR_BANNER)

    def has_error(self) -> bool:
        return self.wait_and_check_visible(_ERROR_BANNER)

File: java-selenium-junit5-cucumber/src/test/java/com/saucedemo/uitests/pages/CheckoutInformationPage.java
package com.saucedemo.uitests.pages;

import com.saucedemo.uitests.testdata.CheckoutInfo;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public final class CheckoutInformationPage extends BasePage {

    private static final By FIRST_NAME_INPUT = By.cssSelector("[data-test='firstName']");
    private static final By LAST_NAME_INPUT = By.cssSelector("[data-test='lastName']");
    private static final By POSTAL_CODE_INPUT = By.cssSelector("[data-test='postalCode']");
    private static final By CONTINUE_BUTTON = By.cssSelector("[data-test='continue']");
    private static final By ERROR_BANNER = By.cssSelector("[data-test='error']");

    public CheckoutInformationPage(WebDriver driver) {
        super(driver);
    }

    public CheckoutOverviewPage fillAndContinue(CheckoutInfo info) {
        typeText(FIRST_NAME_INPUT, info.firstName());
        typeText(LAST_NAME_INPUT, info.lastName());
        typeText(POSTAL_CODE_INPUT, info.postalCode());
        click(CONTINUE_BUTTON);
        return new CheckoutOverviewPage(driver);
    }

    /** Leaves the postal code blank and submits, for the "postal code is required" negative scenario. */
    public void submitWithoutPostalCode(CheckoutInfo info) {
        typeText(FIRST_NAME_INPUT, info.firstName());
        typeText(LAST_NAME_INPUT, info.lastName());
        click(CONTINUE_BUTTON);
    }

    public String errorMessage() {
        return textOf(ERROR_BANNER);
    }

    public boolean hasError() {
        return waitAndCheckVisible(ERROR_BANNER);
    }
}
